### PR TITLE
Documentation for template manager design

### PR DESCRIPTION
**Pull request by WillyPillow**

### CODE REVIEW

File: developer/system/template-manager.md [hunk 1]
```diff
@@ -0,0 +1,259 @@
+---
+layout: doc
+title: Template Manager
+permalink: /doc/template-manager/
+---
+
+Template Manager Design
```
Comment (andrewdavidwong) [inline developer/system/template-manager.md:7]: I think this h1 body title should match the filename and doc index link. I.e., either remove `Design` from this or add it to the others.

File: developer/system/template-manager.md [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+=======================
+
```
Comment (andrewdavidwong) [inline developer/system/template-manager.md:9]: I suggest adding a brief introduction before the first section (`Motivation`).

File: developer/system/template-manager.md [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+Motivation
+----------
+
+Previously, template VMs were distributed by RPM packages and managed by `yum`/`dnf`.
```
Comment (andrewdavidwong) [inline developer/system/template-manager.md:13]: Please use terminology exactly as it appears in the [glossary](https://www.qubes-os.org/doc/glossary/), e.g., `TemplateVM` rather than `template VM`.

File: developer/system/template-manager.md [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+[#2064]: https://github.com/QubesOS/qubes-issues/issues/2064
+
+[#2534]: https://github.com/QubesOS/qubes-issues/issues/2534
+
+[#3040]: https://github.com/QubesOS/qubes-issues/issues/3040
+
+[discussion thread 1]: https://groups.google.com/forum/#!topic/qubes-devel/rwc2_miCNNE/discussion
+
+[discussion thread 2]: https://groups.google.com/forum/#!topic/qubes-users/uQEUpv4THsY/discussion
+
```
Comment (andrewdavidwong) [inline developer/system/template-manager.md:57]: The extra blank lines between these are not necessary. They don't mess anything up, but they take up a lot of space.

File: developer/system/template-manager.md [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+Metadata Storage
+----------------
+
+The template manager needs to keep metadata of installed templates such as versions and origin.
+This data can be stored via `qvm-features` to keep things consistent when, e.g., `qvm-remove` is used.
+Besides, backups are also more easily handled this way.
+
+Also, the fields can serve as an indicator of whether a template is installed by `qvm-template`.
+
+### Fields ###
```
Comment (andrewdavidwong) [inline developer/system/template-manager.md:140]: Minor: It's a bit nicer to use the same heading syntax throughout, but this still works.

Comment (WillyPillow) [reply developer/system/template-manager.md:140]: The reason I did this was that my interpretation of [the guidelines] was to use [markdownlint] `setext_with_atx_closed` (i.e., setext for h1 and h2; closed atx for h3 and above). Perhaps I'm understanding it incorrectly?

[the guidelines]: https://www.qubes-os.org/doc/doc-guidelines/#markdown-conventions
[markdownlint]: https://github.com/DavidAnson/markdownlint/blob/main/doc/Rules.md#md003---heading-style

File: developer/system/template-manager.md (hunk 1 above)
Comment (WillyPillow) [reply developer/system/template-manager.md]: Done (removed `Design` from the title).

File: developer/system/template-manager.md (hunk 3 above)
Comment (WillyPillow) [reply developer/system/template-manager.md]: Fixed.

Just to make sure, is the usage of "template" throughout the document okay?

File: developer/system/template-manager.md (hunk 2 above)
Comment (WillyPillow) [reply developer/system/template-manager.md:15]: Done.

File: developer/system/template-manager.md (hunk 3 above)
Comment (andrewdavidwong) [reply developer/system/template-manager.md]: Yes, as explained in the glossary, that's fine.

File: developer/system/template-manager.md (hunk 5 above)
Comment (andrewdavidwong) [reply developer/system/template-manager.md:127]: Yes. That bullet point is poorly-written. I will fix it.

You want to be consistent with your heading style throughout the document. If you only ever use h1 and h2, then you can simply use setext. However, since you use h3, you must use atx.

Again, this is a minor point, because it will still render correctly. It's just about making the source more consistent.

Comment (andrewdavidwong) [reply developer/system/template-manager.md:127]: On further reflection, it would probably be better for us just to recommend atx universally.

File: developer/system/template-manager.md (hunk 2 above)
Comment (andrewdavidwong) [reply developer/system/template-manager.md:12]: Thanks! I removed the `Introduction` heading, because (1) I believe it is superfluous, and (2) part of my intention (which I failed to make explicit) was to have some body text between the h1 and the first h2, since otherwise it looks rather bad.

Comment (andrewdavidwong) [reply developer/system/template-manager.md:12]: > a template manager application proposed in the 2020 Google Summer of Code program.

Is this intended to be a temporary bit of text that will be removed later? Inevitably, people will read this in the future, then ask us, "When will the Template Manager be updated and ready for real-world use? The documentation says it's  just some GSoC project from back in 2020!"

Comment (WillyPillow) [reply developer/system/template-manager.md:12]: I see. My intent was simply to mention the historical context in which the project was introduced, but I get the concern.

Perhaps it can either be changed to "originally proposed" or be removed outright?

Comment (andrewdavidwong) [reply developer/system/template-manager.md:12]: The `Motivation` section seems like a better place for it.

**Comment by WillyPillow** [comment]
Oops, messed up the rebase, fixing...

File: developer/system/template-manager.md (hunk 2 above)
Comment (WillyPillow) [reply developer/system/template-manager.md:8]: Fixed.

> Is this intended to be a temporary bit of text that will be removed later?

Speaking of this, should there be a piece of text that indicates that the application is not yet merged / does not exist in the stable release of Qubes?

File: developer/system/template-manager.md (hunk 5 above)
Comment (WillyPillow) [reply developer/system/template-manager.md]: The headings should now all be in atx.

File: developer/system/template-manager.md (hunk 2 above)
Comment (andrewdavidwong) [reply developer/system/template-manager.md:8]: > Speaking of this, should there be a piece of text that indicates that the application is not yet merged / does not exist in the stable release of Qubes?

Yes, definitely.

**Comment by andrewdavidwong** [comment]
@marmarta, would you like to review this?

File: developer/system/template-manager.md [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+## Features
+
+- Install/reinstall/downgrade/upgrade templates, either from local packages or remote repositories
+  - Ability to install templates in alternative pools
+  - Possibility for the template package to specify options such as the kernel or `virt_mode` used
+- List and show information about local and available templates
+  - Machine-readable output for easy extension
+- Search for templates
+- Remove templates
+  - Ability to disassociate or remove VMs based on the template to be removed
```
Comment (marmarta) [inline developer/system/template-manager.md:56]: I think it would be nice to clarify what 'disassociate' means here - even a brief phrase like (change their template to another/default template) or something like that would help explain the idea.

File: developer/system/template-manager.md [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+## Package Format
+
+The RPM package format is still used.
+However, the contents are manually extracted instead of installing the whole package.
+This allows us to take advantage of existing tools for things like repository management.
+We can also avoid the burden of dealing with verification, reducing the risk of issues like [QSB-28].
+
+The package name should be in the form `qubes-template-${template_name}`.
+
+For the package metadata (e.g., summary, description), one needs to ensure that the character `|` does not appear.
```
Comment (marmarta) [inline developer/system/template-manager.md:72]: I'd rewrite this for clarity, something like 'In the package metadata (summary, description etc.) should not contain the `|` character to avoid various errors (it is used as an internal separator).'

File: developer/system/template-manager.md [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+    (Notably, the cpio format does not support files over 4GiB.)
+- `var/lib/qubes/vm-templates/${template_name}/template.conf`
+  - Stores custom package metadata (as RPM does not support custom attributes).
+  - Uses `KEY=VALUE` format.
+  - Fields (corresponding to [qvm-prefs]/[qvm-features] tags with the same names)
+    - `virt_mode`
+      - Setting this to `pv` requires user confirmation.
+    - `kernel`
+      - Only allowed to be set to "", i.e., "none", for PVGrub.
+    - `no-monitor-layout`
```
Comment (marmarta) [inline developer/system/template-manager.md:92]: I would rephrase that, someone who is just scanning the list to find something may not notice that five lines below there's a description of those fields. Maybe something like:
- six boolean flags (permitted values: `1` and `0`, corresponding to true and false respectively): `no-monitor-layout`, `pci-e820-host`, `linux-stubdom`, `gui`, `gui-emulated`, `qrexec`

File: developer/system/template-manager.md [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+- `template-url`
+- `template-summary`
+- `template-description`
+  - Note that the newlines in this field are converted to `|` to work better with existing tools like `qvm-features`.
+- `menu-items`
+- `default-menu-items`
+- `netvm-menu-items`
+  - The `*menu-items` entries store the contents of `var/lib/qubes/vm-templates/${template_name}/whitelisted-appmenus.list`, `var/lib/qubes/vm-templates/${template_name}/vm-whitelisted-appmenus.list`, `var/lib/qubes/vm-templates/${template_name}/netvm-whitelisted-appmenus.list` respectively.
+  - Note that newlines are converted to spaces, again for it to work better with existing tools.
+    This should not cause ambiguity as [the FreeDesktop specifications][freedesktop-spec] forbid spaces from the names of desktop files.
```
Comment (marmarta) [inline developer/system/template-manager.md:147]: small rephrasing: forbids spaces in .desktop file names

File: developer/system/template-manager.md [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+## Machine-readable Output
+
+The commands `qvm-template list` and `qvm-template info` provide machine-readable output in both pipe(`|`)-separated and JSON format.
+See the `qvm-template` man page for details.
+
+## Interactions with Existing Tools
+
+### `qvm-remove`
+
+Should work straightforwardly.
```
Comment (marmarta) [inline developer/system/template-manager.md:218]: It's not clear to me whether this means "qvm-remove ignores all this" all "qvm-remove will remove templates properly, not leaving any gunk around" :) I think this needs a bit more explanation.

**Review by marmarta — CHANGES_REQUESTED** [review]
I've added a couple comments, but all in all, great work!

**Comment by andrewdavidwong** [comment]
Still working on this?

**Comment by WillyPillow** [comment]
> Still working on this?

The project, yes, though I unfortunately didn't have a lot of time in the past few weeks due to ICPC and midterms. Now that those things are mostly finished, I should be able to get back on track soon :)

The documentation, OTOH, should be mostly done. (At least no changes are currently planned.) Will go over the conversations above and close resolved ones later.

**Comment by WillyPillow** [comment]
I added some minor changes reflecting later commits. With that done, I think this is mostly complete.

**Comment by andrewdavidwong** [comment]
@WillyPillow, could you please resolve the merge conflict?

**Comment by WillyPillow** [comment]
Done. Also created a PR to update the index at qubesos.github.io.

**Comment by andrewdavidwong** [comment]
> Done. Also created a PR to update the index at qubesos.github.io.

Thanks! Reassigning to @marmarta for final review.

**Comment by andrewdavidwong** [comment]
> > Done. Also created a PR to update the index at qubesos.github.io.
> 
> Thanks! Reassigning to @marmarta for final review.

Oh, GitHub won't let me. @marmarta, consider yourself assigned! :slightly_smiling_face:

**Comment by marmarek** [comment]
PipelineRetry

File: developer/system/template-manager.md [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+[#2534]: https://github.com/QubesOS/qubes-issues/issues/2534
+[#3040]: https://github.com/QubesOS/qubes-issues/issues/3040
+[discussion thread 1]: https://groups.google.com/forum/#!topic/qubes-devel/rwc2_miCNNE/discussion
+[discussion thread 2]: https://groups.google.com/forum/#!topic/qubes-users/uQEUpv4THsY/discussion
+
+## Features
+
+- Install/reinstall/downgrade/upgrade templates, either from local packages or remote repositories
+  - Ability to install templates in alternative pools
+  - Possibility for the template package to specify options such as the kernel or `virt_mode` used
```
Comment (marmarta) [inline developer/system/template-manager.md:51]:
```suggestion
  - Possibility for the template package to specify options such as kernel or `virt_mode` used by the resulting template VM
```

**Comment by marmarta** [comment]
looks good to me, I proposed a minor expansion of one sentence but otherwise looks good

**Comment by andrewdavidwong** [comment]
@marmarta, would you like to add a signed approval comment as described [here](https://www.qubes-os.org/doc/how-to-edit-the-documentation/#security)? @marmarek regularly does this when approving PRs. Perhaps he could help with setting up scripts to do it easily?

**Review by andrewdavidwong — APPROVED** [review]
